[CI/Build]: Add project metadata to pyproject toml file

hickeyma opened the pull request:
The [pyproject.toml](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/) is the more standard way now to handle the project metadata.  This PR moves the metadata from `setup.py` to the `pyproject.toml` file.
It also uses the new [pep 639](https://peps.python.org/pep-0639/) fields for [license and license-files](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license-and-license-files). This requires `setuptools >= 77.0.3`. 

Finally, [dynamic versioning](https://setuptools-scm.readthedocs.io/en/latest/) from `setuptools_scm` is added . This removes the need to manage a static version file as the version file is generated on the fly using [git tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging).

~**Note:** Do not merge until `v0.2.11` tag is removed or updated. This tag is not being recognized by `setuptools-scm` because it is not verified.~ Resolved, thanks @sammshen

hickeyma committed:
Add project metadata to pyproject toml

The pyproject.toml file can handle most the project metadata
in a more central location to the `setup.py` file. This commit
moves it to the toml file.

It also adds dynamic versioning which uses git tags.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update files after ruff check

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sammshen commented on pyproject.toml:
please noete that the version file is just lmcache/__init__.py now. it used to be lmcache/_version.py.

hickeyma replied:
@sammshen Are you saying that previously ye used dynamic versioning or you had `_version.py` static file previously?

sammshen replied:
the latter. I just realized that the _version.py is dynamically generated

sammshen commented:
FYI: your unit test is failing becuase `pip install -e .` failed with the new setup.

```text
+ pip install -e . --no-build-isolation
--
  | Obtaining file:///local/yihua98/.buildkite-agent/builds/nature-1/lmcache/lmcache-unittests
  | Checking if build backend supports build_editable ... done
  | Preparing editable metadata (pyproject.toml) ... error
  | error: subprocess-exited-with-error
  |  
  | × Preparing editable metadata (pyproject.toml) did not run successfully.
  | │ exit code: 1
  | ╰─> [88 lines of output]
  | Building CUDA extensions
  | configuration error: `project.license` must be valid exactly by one definition (2 matches found):
  |  
  | - keys:
  | 'file': {type: string}
  | required: ['file']
  | - keys:
  | 'text': {type: string}
  | required: ['text']
  |  
  | DESCRIPTION:
  | `Project license <https://peps.python.org/pep-0621/#license>`_.
  |  
  | GIVEN VALUE:
  | "Apache-2.0"
  |  
  | OFFENDING RULE: 'oneOf'
  |  
  | DEFINITION:
  | {
  | "oneOf": [
  | {
  | "properties": {
  | "file": {
  | "type": "string",
  | "$$description": [
  | "Relative path to the file (UTF-8) which contains the license for the",
  | "project."
  | ]
  | }
  | },
  | "required": [
  | "file"
  | ]
  | },
  | {
  | "properties": {
  | "text": {
  | "type": "string",
  | "$$description": [
  | "The license of the project whose meaning is that of the",
  | "`License field from the core metadata",
  | "<https://packaging.python.org/specifications/core-metadata/#license>`_."
  | ]
  | }
  | },
  | "required": [
  | "text"
  | ]
  | }
  | ]
  | }
  | Traceback (most recent call last):
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
  | main()
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
  | json_out['return_val'] = hook(**hook_input['kwargs'])
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 181, in prepare_metadata_for_build_editable
  | return hook(metadata_directory, config_settings)
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/build_meta.py", line 469, in prepare_metadata_for_build_editable
  | return self.prepare_metadata_for_build_wheel(
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/build_meta.py", line 373, in prepare_metadata_for_build_wheel
  | self.run_setup()
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/build_meta.py", line 318, in run_setup
  | exec(code, locals())
  | File "<string>", line 59, in <module>
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/__init__.py", line 117, in setup
  | return distutils.core.setup(**attrs)
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 157, in setup
  | dist.parse_config_files()
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/dist.py", line 608, in parse_config_files
  | pyprojecttoml.apply_configuration(self, filename, ignore_option_errors)
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py", line 71, in apply_configuration
  | config = read_configuration(filepath, True, ignore_option_errors, dist)
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py", line 139, in read_configuration
  | validate(subset, filepath)
  | File "/dataheart/yihua98/Applications/anaconda3/envs/buildkite/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py", line 60, in validate
  | raise ValueError(f"{error}\n{summary}") from None
  | ValueError: invalid pyproject.toml config: `project.license`.
  | configuration error: `project.license` must be valid exactly by one definition (2 matches found):
  |  
  | - keys:
  | 'file': {type: string}
  | required: ['file']
  | - keys:
  | 'text': {type: string}
  | required: ['text']
  |  
  | [end of output]
  |  
  | note: This error originates from a subprocess, and is likely not a problem with pip.
  | error: metadata-generation-failed
  |  
  | × Encountered error while generating package metadata.
  | ╰─> See above for output.
  |  
  | note: This is an issue with the package mentioned above, not pip.
  | hint: See above for details.
  | 🚨 Error: The command exited with status 1
  | ^^^ +++
  | user command error: exit status 1


```

sammshen commented:
Just deleted teh v0.2.11 tag

sammshen commented on pyproject.toml:
```diff
@@ -7,14 +7,54 @@ requires = [
+[project]
+name = "lmcache"
+authors = [{name = "LMCache Team", email = "lmcacheteam@gmail.com"}]
+license = "Apache-2.0"
```
license = { text = "Apache-2.0" }

is the fix I believe?

hickeyma replied:
@sammshen Its building ok on my system. I think it is because the build system is using a version of `setuptools` which is < `77.0.3`. This version is where it introduced the new [pep 639](https://peps.python.org/pep-0639/) fields for `license` and `license-files`: https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license-and-license-files

hickeyma replied:
I'll need to add `setuptools` to the reqs file to overcome this.

hickeyma replied:
Fixed now.

sammshen commented:
sorry, one more thing is would you mind adding a small change to the bottom of `.github/PULL_REQUEST_TEMPLATE` that updates the instructions on craeting a new stable release. thank you!

hickeyma committed:
Fix pyproject toml file after rebase

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add setuptools to requirements

This is neded by the build system as it may use any version
of setuptools whihc may not be compatible.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update PR template to remove old tagging information

Now using dynamic versioning, no need anymore to add tags.
This only needs to be done now at release time.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
> sorry, one more thing is would you mind adding a small change to the bottom of .github/PULL_REQUEST_TEMPLATE that updates the instructions on craeting a new stable release. thank you!

Sure, done.

hickeyma commented:
Thanks @sammshen for your reviews and feedback. Updated and ready for review again.

sammshen commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -36,12 +36,4 @@ FIX #xxxx (*link existing issues this PR will resolve*)
 
 <h3>What to Expect for the Reviews</h3>
 
```
IIUC, the new flow to create a stable release will simply be:
```bash
git tag va.b.c
git push origin va.b.c
```
Is this correct?

sammshen replied:
`.github/workflows/publish.yaml` is not edited so I can't leave a comment but the specific section is:
```yaml
      - name: Verify tag matches setup.py version
        run: |
          TAG_VERSION="${GITHUB_REF#refs/tags/v}"
          SETUP_VERSION=$(grep '__version__' lmcache/_version.py | cut -d'"' -f2)
          if [ "$TAG_VERSION" != "$SETUP_VERSION" ]; then
            echo "Git tag version ($TAG_VERSION) does not match setup.py version ($SETUP_VERSION)"
            exit 1
          fi
```
Is the dynamic versioning compatible with this? Sorry I just want to double check.

sammshen replied:
Otherwise, LGTM! :)

hickeyma replied:
Thanks @sammshen for catching the above things. Updated.

hickeyma committed:
Update after review

Review comments:
- https://github.com/LMCache/LMCache/pull/592#pullrequestreview-2823447030

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

YaoJiayi commented on tests/test_cache_engine.py:
```diff
@@ -40,8 +40,8 @@ def generate_tokens(num_tokens, device):
     ret = []
-    for kv_layer in zip(*kv_chunks):
-        klist, vlist = zip(*kv_layer)
+    for kv_layer in zip(*kv_chunks, strict=False):
```
Why adding these `strict=False`?

hickeyma replied:
Not sure why it didn't do this previously but `ruff` linting started to complain as follows:

```command
$ ruff check tests/test_cache_engine.py

tests/test_cache_engine.py:43:21: B905 [*] `zip()` without an explicit `strict=` parameter
   |
41 |     dim = 1 if fmt == "huggingface" else 0
42 |     ret = []
43 |     for kv_layer in zip(*kv_chunks):
   |                     ^^^^^^^^^^^^^^^ B905
44 |         klist, vlist = zip(*kv_layer, strict=False)
45 |         klayer = torch.cat(klist, dim=dim)
   |
   = help: Add explicit `strict=False`

Found 1 error.
[*] 1 fixable with the `--fix` option.
```

YaoJiayi replied:
What's your ruff version?

hickeyma replied:
`0.6.5`, same as the reqs file.

YaoJiayi replied:
Hmm. So weird. Let me try to replicate real quick.

YaoJiayi replied:
On my side, it's `All checks passed!` when running `ruff check tests/test_cache_engine.py`.

hickeyma replied:
Did you pull down this PR and run lint with it?

YaoJiayi reviewed (commented):
LGTM in general. Just curious why these `strict=False` are added:)

hickeyma commented:
> LGTM in general. Just curious why these strict=False are added:)

@YaoJiayi Thank you for the review. Responded to the above in https://github.com/LMCache/LMCache/pull/592#discussion_r2084221740

YaoJiayi reviewed (approved):
LGTM! Thanks @hickeyma
